modules/wallet/update.go

```diff
@@ -10,21 +10,93 @@ import (
 	"github.com/NebulousLabs/bolt"
 )
 
-// historicOutput defines a historic output as recognized by the wallet. This
-// struct is primarily used to sort the historic outputs before inserting them
-// into the bolt database.
-type historicOutput struct {
-	id  types.OutputID
-	val types.Currency
+// threadedResetSubscriptions unsubscribes the wallet from the consensus set and transaction pool
+// and subscribes again.
+func (w *Wallet) threadedResetSubscriptions() error {
+	if !w.scanLock.TryLock() {
```

Comment:
Looks like both the update functions are missing threadgroup interactions as well.

Right now, if you close the wallet, this will be left running in the background. If the wallet is shutting down correctly, it'll close the log, which means any action which gets logged will cause a panic if there's a subscription running in the background.